design/arp-load-sharing.md

```diff
@@ -0,0 +1,41 @@
+# ARP Load Sharing
+
+## Summary
+
+MetalLB in layer 2 mode does not support load balancing. See [here](https://metallb.universe.tf/concepts/layer2/).
+However, in some cases, L2 is the only option, yet we still require a load balancing solution which is internal to our infrastructure.
+
+A method that can achieve that is known as "ARP Load Sharing". See for example [here](https://docs.paloaltonetworks.com/pan-os/10-0/pan-os-admin/high-availability/ha-concepts/arp-load-sharing). This technique involves fiddling with how ARP replies are sent, which raises the concern that it might get blocked by the network switch or security policy since it seems like an [ARP Spoofing](https://en.wikipedia.org/wiki/ARP_spoofing) attack. See more information about "does arp spoofing work on all lans" [here](https://security.stackexchange.com/questions/133784/does-arp-spoofing-work-on-all-lans).
+
+This document proposes adding an option to MetalLB to support "ARP Load Sharing" mode.
```

Comment:
Is this only expected to be useful when multiple clients are on the same layer2 domain as the cluster?  Otherwise, the client issuing the ARP request is just the router, right?

I see in the palo alto doc, it lists this limitation: "Use ARP load-sharing only when no Layer 3 device exists between the firewall and end hosts, that is, when end hosts use the firewall as their default gateway."

Does this match your use case?

Reply:
@russellb Thanks! that's a good point.

A router will appear as a **single** L2 client to the cluster, so this is indeed a limitation of this mode for clients behind a router. However for clients on the same LAN/VLAN this approach provides the best balancing L2 can buy.

My use-case is an object-store software, that runs on all the cluster nodes, so even the relatively simple clients balancing over nodes allows to scale to the full throughput of the cluster's network.

I'm pretty sure that this mode should be considered experimental until proven otherwise, it might get routers to block the cluster ports. But it will look like magic in the hands of a skilled admin.

Reply:
Yeah, it's pretty clever for the environments where it would work.  I would definitely agree that it should be experimental.

So then we're left deciding whether to take this experimental mode and the added complexity it introduces, for a limited use case (only useful to network-local clients, and potentially problematic in some environments).  With my maintainer hat on, I think I'd want to see some more demand from people saying they would use this mode.

With my Red Hat on, I don't think I'd want to support this.  I'm pretty nervous of it resulting in expensive support engagements figuring out why someone's network is not behaving as we expect with this mode.